src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs

```diff
@@ -0,0 +1,131 @@
+            {
+                ThrowHelper.ThrowJsonException_MetadataValueWasNotString();
+            }
+
+            MetadataPropertyName metadata = state.Current.MetadataProperty;
+            string key = reader.GetString()!;
+
+            if (metadata == MetadataPropertyName.Id)
+            {
+                state.ReferenceResolver.AddReference(key!, GetValueToPreserve(ref state));
+            }
+            else if (metadata == MetadataPropertyName.Ref)
+            {
+                state.Current.ReferenceId = key;
+            }
+
+            state.Current.ReadMetadataValue = false;
+        }
+
+        private static object GetValueToPreserve(ref ReadStack state)
+        {
+            return state.Current.IsProcessingProperty(ClassType.Dictionary) ?
+                state.Current.JsonPropertyInfo!.GetValueAsObject(state.Current.ReturnValue)! : state.Current.ReturnValue!;
+        }
+
+        internal static MetadataPropertyName GetMetadataPropertyName(ReadOnlySpan<byte> propertyName, ref ReadStack state, ref Utf8JsonReader reader)
+        {
+            if (propertyName[0] == '$')
+            {
+                switch (propertyName.Length)
+                {
+                    case 3:
+                        if (propertyName[1] == 'i' &&
+                            propertyName[2] == 'd')
+                        {
+                            return MetadataPropertyName.Id;
+                        }
+                        break;
+
+                    case 4:
+                        if (propertyName[1] == 'r' &&
+                            propertyName[2] == 'e' &&
+                            propertyName[3] == 'f')
+                        {
+                            return MetadataPropertyName.Ref;
+                        }
+                        break;
+
+                    case 7:
+                        // Only Preserved Arrays are allowed to read $values as metadata.
+                        if (state.Current.IsPreservedArray &&
+                            propertyName[1] == 'v' &&
+                            propertyName[2] == 'a' &&
+                            propertyName[3] == 'l' &&
+                            propertyName[4] == 'u' &&
+                            propertyName[5] == 'e' &&
+                            propertyName[6] == 's')
+                        {
+                            return MetadataPropertyName.Values;
+                        }
+                        break;
+                }
+
+                // Fail state.
+                // Set PropertyInfo or KeyName to write down the conflicting property name in JsonException.Path
+                if (state.Current.IsProcessingDictionary())
+                {
+                    state.Current.KeyName = reader.GetString();
+                }
+                else
+                {
+                    JsonPropertyInfo info = JsonPropertyInfo.s_metadataProperty;
```

Comment:
I believe based on my other comments, you'll need to add another property to readstate for the name. FWIW I added that property (for a different reason) in a private refactoring branch called "JsonPropertyNameAsString" and updated the JsonPath code to inspect it. Perhaps just remove this code that uses JsonPropertyInfo and add a TODO and I'll hook it up when I merge in the refactoring branch.